features/cellular/framework/mux/cellular_mux.cpp

```diff
@@ -0,0 +1,1253 @@
+        size = max_payload_size;
+    }
+
+    _mutex.lock();
+
+    ssize_t write_ret;
+    switch (_tx_context.tx_state) {
+        case TX_IDLE:
+            if (!_state.is_tx_callback_context) {
+                /* Proper state to start TX cycle within current context. */
+
+                user_information_construct(dlci_id, buffer, size);
+                tx_state_change(TX_NORETRANSMIT, tx_noretransmit_entry_run, tx_idle_exit_run);
+
+                write_ret = size;
+            } else {
+                /* Current context is TX callback context. */
+
+                if (!_state.is_user_tx_pending) {
+                    /* Signal callback context to start TX cycle and construct the frame. */
+
+                    /* @note: If user is starting a TX to a DLCI, which is after the current DLCI TX callback within the
+                     * stored sequence this will result to dispatching 1 unnecessary TX callback, if this is a issue one
+                     * should clear the TX callback pending bit marker for this DLCI in this place. */
+
+                    _state.is_user_tx_pending = 1u;
+                    user_information_construct(dlci_id, buffer, size);
+
+                    write_ret = size;
+                } else {
+                    /* TX cycle allready scheduled, set TX callback pending and inform caller by return value that no
+                       action was taken. */
+
+                    tx_callback_pending_bit_set(dlci_id);
+                    write_ret = 0;
+                }
+            }
+
+            break;
+        default:
+            /* TX allready in use, set TX callback pending and inform caller by return value that no action was taken.
+             */
+            tx_callback_pending_bit_set(dlci_id);
+            write_ret = 0;
+
+            break;
+    }
+
+    _mutex.unlock();
+
+    return write_ret;
+}
+
+
+size_t Mux::min(uint8_t size_1, size_t size_2)
+{
+    return (size_1 < size_2) ? size_1 : size_2;
+}
+
+
+ssize_t Mux::user_data_rx(void* buffer, size_t size)
+{
+    MBED_ASSERT(buffer != NULL);
+
+    _mutex.lock();
+
+    if (_state.is_user_rx_ready) {
+        const size_t read_length = min((_rx_context.read_length - _rx_context.offset), size);
+        memcpy(buffer, &(_rx_context.buffer[FRAME_INFORMATION_FIELD_INDEX + _rx_context.offset]), read_length);
+        _rx_context.offset += read_length;
+
+        if (_rx_context.offset == _rx_context.read_length) {
+            /* Current Rx buffer consumed, disable user reading and resume the Rx path. */
+
+            _state.is_user_rx_ready = 0;
+            rx_event_do(RX_RESUME);
+        }
+
+        _mutex.unlock();
+
+        return static_cast<ssize_t>(read_length);
+    } else {
+        _mutex.unlock();
+
+        return -EAGAIN;
+    }
+}
+
+
+short Mux::poll()
+{
+    _mutex.lock();
+
+    const bool writable = (_tx_context.tx_state == TX_IDLE);
+    const bool readable = (_rx_context.rx_state == RX_SUSPEND);
```

Comment:
This looks correct for writable, but I'm not sure about the readable. I think the condition should be `is_user_ready` - this should be the indication "read will now give you something, not just say -EAGAIN".

Reply:
Using _state.is_user_rx_ready should be also fine but prefer doing it with like this at it maps to design description:
http://trac.oulu.arm.com/trac/nanoservice/wiki/NBIoT/Mux
- ch Multiplexer frame RX statemachine - BASIC MODE

As thinking now _state.is_user_rx_ready could be removed completely as RX SM should have all required information available => tracked in IOTCELL-584

Reply:
I'm not famililar with the structure, but the thing that prompted the comment was the fact the "I'm ready" condition is apparently different to that in the user_data_rx routine that implements read() above. Whichever you prefer, you should align the routines.